app/assets/src/components/views/PipelineViz.jsx

```diff
@@ -0,0 +1,366 @@
+      }
+    });
+
+    // Add beginning (input) and ending (output) nodes
+    nodeData.push({ id: START_NODE_ID, level: 0, group: "startEndNodes" });
+    nodeData.push({ id: END_NODE_ID, group: "startEndNodes" });
+
+    let maxLevel = 1;
+    stepData.forEach((step, i) => {
+      step.in.forEach(inTarget => {
+        if (inTarget in outTargetToStepId) {
+          const fromId = outTargetToStepId[inTarget];
+          edgeData.push({ from: fromId, to: i });
+
+          nodeData[i].level = Math.max(
+            nodeData[i].level,
+            nodeData[fromId].level + 1
+          );
+          maxLevel = Math.max(maxLevel, nodeData[i].level);
+        } else {
+          // Connect beginning steps to input node
+          edgeData.push({ from: START_NODE_ID, to: i });
+        }
+      });
+    });
+    nodeData[stepData.length + 1].level = maxLevel + 1;
+  }
+
+  adjustGraphNodePositions() {
+    let centerDOMCoords;
+    this.stageGraphsInfo.forEach((graphInfo, i) => {
+      const graph = graphInfo.graph;
+      this.adjustStageWidth(graph);
+      if (i == 0) {
+        const canvasCoords = graph.getPositions([START_NODE_ID])[START_NODE_ID];
+        centerDOMCoords = graph.canvasToDOM(canvasCoords);
+      } else {
+        const canvasYCoord = graph.DOMtoCanvas(centerDOMCoords).y;
+        graph.moveNode(
+          START_NODE_ID,
+          graph.getPositions([START_NODE_ID])[START_NODE_ID].x,
+          canvasYCoord
+        );
+        graph.moveNode(
+          END_NODE_ID,
+          graph.getPositions([END_NODE_ID])[END_NODE_ID].x,
+          canvasYCoord
+        );
+      }
+    });
+  }
+
+  adjustStageWidth(graph) {
+    // Set initial zoom for width calculation.
+    graph.moveTo({ scale: 1 });
+
+    // Calculate and set new canvas width.
+    const minX = graph.canvasToDOM({
+      x: graph.getBoundingBox(START_NODE_ID).right,
+    }).x;
+    // Use right instead of left to include room for edge arrow tip.
+    const maxX = graph.canvasToDOM({
+      x: graph.getBoundingBox(END_NODE_ID).right,
+    }).x;
+    graph.setSize(maxX - minX + "px", "100%");
+
+    // Reset zoom (which is adjusted in setSize).
+    graph.moveTo({ scale: 1, position: { x: 0, y: 0 } });
+  }
+
+  renderGraphs() {
+    this.modifyStepNames();
+    this.stageNames.forEach((stageName, i) => {
+      this.renderStageGraph(
+        this.stageResults[stageName],
+        this[`container${i}`],
+        i
+      );
+    });
+    this.renderInterStageEdges();
+    this.adjustGraphNodePositions();
+  }
+
+  renderInterStageEdges() {
+    this.stageGraphsInfo.forEach((currStageInfo, i) => {
+      if (i == this.stageGraphsInfo.length - 1) {
+        // Create hidden edges to final node for vertical centering of nodes.
+        currStageInfo.data.nodes.forEach(nodeData => {
+          currStageInfo.data.edges.add({
+            from: nodeData.id,
+            to: END_NODE_ID,
+            color: {
+              color: STAGE_BG_COLOR,
```

Comment:
you may want to consider some sort of className based coloring instead of directly setting styles here